BIP short

amal-ghamdi opened the pull request:
Thank you @chaozg for reviewing this, the compiled version can be seen at this test book https://amal-ghamdi.github.io/CUQI-Book-test/intro.html

closes #63

chaozg commented on chapter01/Intro_example_short.ipynb:
```diff
@@ -0,0 +1,671 @@
+   "source": [
+    "Consider the following inverse problem: given observed data $b$, determine $x_1$, and $x_2$:\n",
+    "\n",
+    "$$b = x_1 + x_2 + e \\;\\;\\mathrm{with}\\;\\; e \\sim \\mathrm{Gaussain}(0, 0.1)$$\n",
```
```suggestion
    "$$b = x_1 + x_2 + e \\;\\;\\mathrm{with}\\;\\; e \\sim \\mathrm{Gaussian}(0, 0.1)$$\n",
```

chaozg commented on chapter01/Intro_example_short.ipynb:
```diff
@@ -0,0 +1,671 @@
+    "## <font color=#CD853F> 0. Learning objectives </font> <a name=\"r-learning-objectives\"></a>\n",
+    "  * Create a simple BIP in CUQIpy\n",
+    "      * Create a linear forward model object\n",
+    "      * Create distribution objects that represent the prior, the noise, and the data distributions\n",
```
data "distribution" instead of "distributions"?

amal-ghamdi replied:
the word "distributions" is meant for prior, noise, and data as in prior distribution, noise distribution, and data distribution.

chaozg commented on chapter01/Intro_example_short.ipynb:
```diff
@@ -0,0 +1,671 @@
+    "\n",
+    "$$b = x_1 + x_2 + e \\;\\;\\mathrm{with}\\;\\; e \\sim \\mathrm{Gaussain}(0, 0.1)$$\n",
+    "\n",
+    "We can write it as:\n",
```
```suggestion
    "We can also write it in the following matrix form:\n",
```

chaozg commented on chapter01/Intro_example_short.ipynb:
```diff
@@ -0,0 +1,671 @@
+    "|$e$          |noise                    |1-dimensional  ​|\n",
+    "\n",
+    "This problem is:\n",
+    "* Considered a Linear inverse problem since the forward model is linear.\n",
```
```suggestion
    "* Considered a linear inverse problem since the forward model is linear.\n",
```

amal-ghamdi replied:
I think both writing seems alright, I am leaning toward the first one :)

chaozg replied:
What about the L in "Linear"? I guess you capitalize it for emphasis?

amal-ghamdi replied:
Got it, I did not pay attention to the capitalization. I fixed it. Thanks!

chaozg commented on chapter01/Intro_example_short.ipynb:
```diff
@@ -0,0 +1,671 @@
+    "\n",
+    "This problem is:\n",
+    "* Considered a Linear inverse problem since the forward model is linear.\n",
+    "* Ill-posed (in the sense of Hadamard [1]) since the solution is not unique, i.e., for some given value of $b$, e.g., $b=3$, all points $(x_1, x_2)$ that satisfy $x_1 + x_1 = 3$ are solutions to the (noise-free) problem."
```
```suggestion
    "* Ill-posed (in the sense of Hadamard [1]) since the solution is not unique, i.e., for some given value of $b$, e.g., $b=3$, all parameter pairs $(x_1, x_2)$ satisfying $x_1 + x_1 = 3$ are solutions to the (noise-free) problem."
```

chaozg commented on chapter01/Intro_example_short.ipynb:
```diff
@@ -0,0 +1,671 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let us define the forward model $\\mathbf{A}$, we first define the matrix:"
```
```suggestion
    "Let us define the forward model $\\mathbf{A}$. We start by specifying the underlying matrix:"
```

chaozg commented on chapter01/Intro_example_short.ipynb:
```diff
@@ -0,0 +1,671 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now we wrap `A_matrix` in a `CUQIpy` forward model object, we name `A`, as follows:"
```
```suggestion
    "Now we wrap `A_matrix` in a `CUQIpy` forward model object, which we name `A`, as follows:"
```

chaozg commented on chapter01/Intro_example_short.ipynb:
```diff
@@ -0,0 +1,671 @@
+    "- $\\delta$ is the standard deviation of the prior distribution,\n",
+    "- $\\mathbf{I}$ is the identity matrix.\n",
+    "\n",
+    "Let us define the prior distribution for the parameters $x_1$ and $x_2$:"
```
Just curious if we want to write x1 and x2 explicitly here or just x. Probably make it clear that we are now creating the prior in CUQIpy or with the CUQIpy Gaussian distribution?

amal-ghamdi replied:
done :)

chaozg commented on chapter01/Intro_example_short.ipynb:
```diff
@@ -0,0 +1,671 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Before sampling or evaluating the PDF of `b`, we need to specify the value of the parameters `x`, let us choose the following value:"
```
```suggestion
    "Before sampling or evaluating the PDF of `b`, we need to specify the value of the parameters`x`. let us choose the following value:"
```

chaozg commented on chapter01/Intro_example_short.ipynb:
```diff
@@ -0,0 +1,671 @@
+    "$$L (\\mathbf{x} | b^\\mathrm{obs}) \\mathrel{\\vcenter{:}}= \\pi (b^\\mathrm{obs} | \\mathbf{x})$$\n",
+    "\n",
+    "\n",
+    "Example, given observed data  $b^\\mathrm{obs} $\n",
```
The sentence here is not complete, something like "Since we are using a Gaussian noise model, the likelihood can be evaluated as"?

amal-ghamdi replied:
agreed :)

chaozg commented on chapter01/Intro_example_short.ipynb:
```diff
@@ -0,0 +1,671 @@
+    "\n",
+    "$$\\pi(\\mathbf{x} | b) \\propto \\pi( b|\\mathbf{x})\\pi(\\mathbf{x})$$\n",
+    "\n",
+    "- Note that $\\pi( b|\\mathbf{x})$ here denotes the likelihood and not the data distribution, despite often written that way.\n",
```
$b$ here and in the equation above is actually $b_{obs}$, right? if we stick to the notation introduced...

amal-ghamdi replied:
you are right, I changed the writing a bit to clarify that

chaozg commented on chapter01/Intro_example_short.ipynb:
```diff
@@ -0,0 +1,671 @@
+    " \n",
+    "- Maximizer of the posterior\n",
+    "\n",
+    "$$\\mathbf{x}^* = \\underset{\\mathbf{x}}{\\operatorname{argmax\\;}} \\pi(\\mathbf{x} | b)$$\n",
```
$b$ or $b_{obs}$...

amal-ghamdi replied:
good point, fixed!

chaozg commented on chapter01/Intro_example_short.ipynb:
```diff
@@ -0,0 +1,671 @@
+   "source": [
+    "### <font color=#8B4513> 2.2. Sampling from the posterior </font> <a class=\"anchor\" id=\"r-sampling-from-the-posterior\"></a> \n",
+    "\n",
+    "The MAP point is a very useful point estimate, but it does not provide information about the uncertainty in the estimate. Thus, we need to sample from the posterior distribution to quantify the uncertainty."
```
```suggestion
    "The MAP point is a very useful point estimate, but it does not provide information about the uncertainty associated with the estimate. To quantify this uncertainty, we need to draw samples from the posterior distribution"
```

chaozg reviewed (approved):
Thanks @amal-ghamdi for creating such a shorter version of the simplest BIP notebook. Well done, and I only have some tiny suggestions on the grammar and comments on the user of $b$ and $b_{obs}$

amal-ghamdi commented:
Thank you @chaozg for the very helpful review! I addressed all your comments and I am requesting a re-review. Feel free to close conversations that you think were addressed.

chaozg reviewed (approved):
LGTM

amal-ghamdi commented:
Thank you @chaozg  for the second round :), I am re-requesting a review

chaozg commented on chapter01/Intro_example_short.ipynb:
```diff
@@ -0,0 +1,669 @@
+    "\n",
+    "Since we are using a Gaussian noise model, the likelihood can be evaluated as\n",
+    "\n",
+    "$$L (x_1, x_2 | b=b^\\mathrm{obs}) = \\frac{1}{\\sqrt{2 \\pi \\cdot 0.1}} \\mathrm{exp}\\left(-\\frac{(x_1+x_2- b^\\mathrm{obs})^2}{2\\cdot 0.1}\\right) $$\n",
```
Do you think it's better to include A in the equation so that the link between a likelihood and a forward operator is more visible?

amal-ghamdi replied:
good point, I added an intermediate form with A, then with x_1 and x_2 to make the link as you suggested and to show the reader the concrete formula.

amal-ghamdi replied:
btw, the test link above in the PR description is updated with these latest changes

amal-ghamdi commented on chapter01/Intro_example_short.ipynb:
```diff
@@ -0,0 +1,669 @@
+    "\n",
+    "\n",
+    "## <font color=#CD853F> Contents of this notebook: </font>\n",
+    "  * [0. Learning objectives](#r-learning-objectives)\n",
```
Thanks! I fixed it now. This is a good point about what we need in terms of the notebook table of contents TOC. To accommodate both browsing from the book and working on VS code, it could be helpful to have the TOC repeated. But it is a bit of overkill. We can maybe put the main sections only at the top of the notebook and not subsections (which is consistent with some of the training material we have, eg, Gibbs notebook). We can talk about today about it

amal-ghamdi replied:
Based on our discussion the other day, I removed subheaders from the `Contents of this notebook` section.

nabriis replied:
I could not get the links to work either in https://amal-ghamdi.github.io/CUQI-Book-test/chapter01/Intro_example_short.html#r-solving-the-bip at the time of writing.

amal-ghamdi replied:
Sorry the test book is now ahead of this branch, eventually all changes in the test book will be added on top of this branch. You can see this notebook in the test book https://amal-ghamdi.github.io/CUQI-Book-test/chapter01/Intro_example_short.html (no changes done to this notebook in particular in the test branch)

amal-ghamdi commented:
Thank you @chaozg  for your comments. I made updates to the TOC, I am also using the utilities plotting methods now. I am re-requesting your review and I appreciate it if you can take a look. Thanks!

nabriis commented on chapter01/Intro_example_short.ipynb:
```diff
@@ -0,0 +1,604 @@
+    "from cuqi.distribution import Gaussian\n",
+    "from cuqi.problem import BayesianProblem\n",
+    "from cuqi.model import LinearModel\n",
+    "from cuqi.utilities import plot_1D_density, plot_2D_density\n",
```
The rendered notebook on the link you provide throws an error here. Unable to import plot_1D_density.

amal-ghamdi replied:
Thanks @nabriis 

You can see this fixed in https://amal-ghamdi.github.io/CUQI-Book-test/chapter01/Intro_example_short.html (in the test book)

I added --pre in the requirements file to use the pre-release and now no such error

nabriis commented on chapter01/Intro_example_short.ipynb:
```diff
@@ -0,0 +1,604 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Note that the likelihood function is a density function and is not a distribution. If we try to compute `pdf` for example, we will get an error:"
```
```suggestion
    "Note that the likelihood function is a density function and is not a distribution. If we try to compute `pdf` for example, we will get an error."
```

nabriis commented on chapter01/Intro_example_short.ipynb:
```diff
@@ -0,0 +1,604 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Note:** One method of estimating the inverse proble solution is to maximize the likelihood function, which is equivalent to minimizing the negative log-likelihood function. This is known as the maximum likelihood (ML) point estimate.\n",
```
```suggestion
    "**Note:** One method of estimating the inverse problem solution is to maximize the likelihood function, which is equivalent to minimizing the negative log-likelihood function. This is known as the maximum likelihood (ML) point estimate.\n",
```

nabriis commented on chapter01/Intro_example_short.ipynb:
```diff
@@ -0,0 +1,604 @@
+    "### <font color=#8B4513> 1.6. Putting it all together, the BIP </font> <a class=\"anchor\" id=\"r-putting-it-all-together-the-bip\"></a>\n",
+    "\n",
+    "#### Posterior definition using the Bayes’ rule\n",
+    "- The posterior is proportional to the product of **likelihood** and **prior**\n",
```
```suggestion
    "The posterior is proportional to the product of **likelihood** and **prior**\n",
```

nabriis commented on chapter01/Intro_example_short.ipynb:
```diff
@@ -0,0 +1,604 @@
+   "source": [
+    "### <font color=#8B4513> 2.1. Maximum a posteriori (MAP) estimate </font> <a class=\"anchor\" id=\"r-maximum-a-posteriori-map-estimate\"></a>\n",
+    " \n",
+    "- Maximizer of the posterior\n",
```
```suggestion
    "Maximizer of the posterior\n",
```

nabriis commented on chapter01/Intro_example_short.ipynb:
```diff
@@ -0,0 +1,604 @@
+    "- Maximizer of the posterior\n",
+    "\n",
+    "$$\\mathbf{x}^* = \\underset{\\mathbf{x}}{\\operatorname{argmax\\;}} \\pi(\\mathbf{x} | b=b^\\mathrm{obs})$$\n",
+    "- In the case with Gaussian noise and Gaussian prior, this is the classic Tikhonov solution\n",
```
```suggestion
    "In the case with Gaussian noise and Gaussian prior, this is the classic Tikhonov solution\n",
```

nabriis commented on chapter01/Intro_example_short.ipynb:
```diff
@@ -0,0 +1,604 @@
+    "plot_2D_density(BP.posterior, x1_lim[0], x1_lim[1], x2_lim[0], x2_lim[1])\n",
+    "\n",
+    "# Plot the posterior samples\n",
+    "samples.plot_pair(ax=plt.gca())\n",
```
Nice :)

nabriis reviewed (approved):
Hi @amal-ghamdi. The notebook looks nice! I added some small suggestions you can consider batch adding. Once you fix also getting the notebook to render (run) on the book page I think it can get merged.

amal-ghamdi commented:
Many thanks, @chaozg and @nabriis, for the review!

I have addressed your comments.

you can see this notebook (including Nicolai's latest update) at https://amal-ghamdi.github.io/CUQI-Book-test/chapter01/Intro_example_short.html (Note: the test book is generated from a different branch that is ahead of this branch but no changes were made to this notebook specifically as far as I know)

chaozg commented on chapter01/Intro_example_short.ipynb:
```diff
@@ -0,0 +1,604 @@
+    " \n",
+    "Maximizer of the posterior\n",
+    "\n",
+    "$$\\mathbf{x}^* = \\underset{\\mathbf{x}}{\\operatorname{argmax\\;}} \\pi(\\mathbf{x} | b=b^\\mathrm{obs})$$\n",
```
There seems a jump between this equation and the one below? What do you think? For example, from maximization to minimization.

amal-ghamdi replied:
Thanks for the MS teams chat on this, resolved based on that chat.

chaozg commented on chapter01/Intro_example_short.ipynb:
```diff
@@ -0,0 +1,604 @@
+   "source": [
+    "### <font color=#8B4513> 2.2. Sampling from the posterior </font> <a class=\"anchor\" id=\"r-sampling-from-the-posterior\"></a> \n",
+    "\n",
+    "The MAP point is a very useful point estimate, but it does not provide information about the uncertainty associated with the estimate. To quantify this uncertainty, we need to draw samples from the posterior distribution"
```
Is it good to put the equation of the posterior distribution here, so the link between a MAP and posterior is more obvious? Just a thought.

chaozg reviewed (approved):
Amal, thanks for your further update, I only left some tiny ideas in the comments.
In addition, it seems the links in toc are still not working, as reported by @nabriis . Of course, this issue is not limited to this notebook and we should probably address them all later in a separate PR.

amal-ghamdi commented:
> Amal, thanks for your further update, I only left some tiny ideas in the comments. In addition, it seems the links in toc are still not working, as reported by @nabriis . Of course, this issue is not limited to this notebook and we should probably address them all later in a separate PR.

Thank you @chaozg for your review. The links in the top of the notebook works for me on VS code but does not work in the online book and I could not find a way to fix it yet. In the online book, the navigation on the right works to move between notebook sections. I created a separate issue to be addressed in the future for this #67.
